### PR TITLE
drivers: can: add ESP32 TWAI CAN controller driver

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Add ESP32 TWAI CAN controller driver. The TWAI is register-compatible with the NXP SJA1000 CAN controller.
    
Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

### CODE REVIEW

File: dts/xtensa/espressif/esp32.dtsi
Comment (galak) [inline dts/xtensa/espressif/esp32.dtsi:283]: remove `label` prop

Comment (henrikbrixandersen) [reply dts/xtensa/espressif/esp32.dtsi]: Removed.

**Review by galak — COMMENTED** [review]
What's the relationship between the ESP32 CAN & SJA1000?

Seems like SJA1000 should have its own DTS binding, etc.

**Comment by henrikbrixandersen** [comment]
> What's the relationship between the ESP32 CAN & SJA1000?

The only relationship is that the Espressif ESP32 TWAI peripheral is register-compatible with the NXP SJA1000.

> Seems like SJA1000 should have its own DTS binding, etc.

The NXP SJA1000 is a stand-alone CAN controller with a legacy MCU interface. If anybody were to add Zephyr support for that as a stand-alone CAN controller, it would require its own DTS binding.

The register layout of the NXP SJA1000, however, is used in several other CAN controllers (e.g. the ESP32 TWAI). It is also used on e.g. the older Kvaser CAN PCI cards, which are supported as an emulated CAN controller under qemu (https://www.qemu.org/docs/master/system/devices/can.html). I plan to use the `sja1000.c` backend driver in several front-end drivers, but I have no intentions of adding support for stand-alone (legacy) SJA1000 controllers.

**Comment by galak** [comment]
> The NXP SJA1000 is a stand-alone CAN controller with a legacy MCU interface. If anybody were to add Zephyr support for that as a stand-alone CAN controller, it would require its own DTS binding.
> 
> The register layout of the NXP SJA1000, however, is used in several other CAN controllers (e.g. the ESP32 TWAI). It is also used on e.g. the older Kvaser CAN PCI cards, which are supported as an emulated CAN controller under qemu (https://www.qemu.org/docs/master/system/devices/can.html). I plan to use the `sja1000.c` backend driver in several front-end drivers, but I have no intentions of adding support for stand-alone (legacy) SJA1000 controllers.

Was thinking about this, and that all makes sense.  Was thinking maybe we rename things a little to `sja1000_lib.c` and the Kconfig to match.  So this is kinda the "library" code to provide the sja1000 funcitonality.

File: drivers/can/can_esp32_twai.c
Comment (uLipe) [inline drivers/can/can_esp32_twai.c:52]: Question, this APB_CLK_FREQ should not be get using some function from clock control device driver?

File: drivers/can/can_esp32_twai.c
Comment (uLipe) [inline drivers/can/can_esp32_twai.c:84]: esp_intr_alloc may fail if heap size is insufficient for all ISR descriptors, I suggest to check the return here and report if it fails.

File: dts/xtensa/espressif/esp32.dtsi
Comment (uLipe) [inline dts/xtensa/espressif/esp32.dtsi:284]: Is this a flexible parameter? The sample point according to CAN bus specification should reside between 60 to 70% of bit rate period and depends on SJW and the CAN BUS time quanta 1 and 2, what would happen here if other bit rate got selected this value will be the same? Same for the sjw, it would assume values between 1 and 3 time quanta.

**Review by uLipe — CHANGES_REQUESTED** [review]
In general LGTM,  minor nits. Just check the comment about the interrupt allocator above, this is the only major thing I think it should be checked.

Also, @sylvioalves and other espessif maintainers may have comments on how implement this driver using the espressif hal instead of direct register manipulation in order to make this driver consistent with the other esp32 device drivers.

File: drivers/can/can_esp32_twai.c
Comment (henrikbrixandersen) [reply drivers/can/can_esp32_twai.c:52]: Ideally, yes. There is no ESP32 clock controller driver allowing to retrieve this clock rate. Other in-tree ESP32 drivers requiring this clock rate also use the define.

File: drivers/can/can_esp32_twai.c
Comment (henrikbrixandersen) [reply drivers/can/can_esp32_twai.c:84]: I cannot find any other in-tree ESP32 driver checking the return value of `esp_intr_alloc()`.

File: dts/xtensa/espressif/esp32.dtsi
Comment (henrikbrixandersen) [reply dts/xtensa/espressif/esp32.dtsi:284]: According to CAN in Automation (CiA) the sample point shall reside at 87.5%. This is just a default setting, though. Both the sample point and the SJW can be changed at runtime using the CAN timing APIs.

**Comment by henrikbrixandersen** [comment]
> Was thinking maybe we rename things a little to `sja1000_lib.c` and the Kconfig to match. So this is kinda the "library" code to provide the sja1000 funcitonality.

I prefer to keep the naming as-is.

**Review by str4t0m — COMMENTED** [review]
Thanks, this looks great.
I'll test it soon, and only have a few comments so far.

File: drivers/can/can_sja1000.c
Comment (martinjaeger) [reply drivers/can/can_sja1000.c:641]: That's really unfortunate that they are not fully compatible. I was about to test the PR on an ESP32-C3, assuming it would only require to add the dts node with the correct register offset.

Comment (henrikbrixandersen) [reply drivers/can/can_sja1000.c:641]: Yes. It is unfortunate. I too discovered these changes between the ESP32 TWAI and those present in other Espressif SoCs.

My plan is handle these by replacing the affected generic SJA1000 API functions with custom ones for the SoCs that do not adhere to the SJA1000 register layout (e.g. using a custom init function, custom timing function). This will allow us to keep non-SJA1000 compliant functionality/extension out of the common backend and isolate it to the given front-end driver.

File: drivers/can/can_sja1000.c
Comment (henrikbrixandersen) [reply drivers/can/can_sja1000.c:685]: Well, reading back the OCR register on the ESP32 TWAI does return a valid setting - but writes seem to be ignored. I am just setting it to whatever the TWAI seems to be hardcoded to use.

File: drivers/can/can_sja1000.c
Comment (henrikbrixandersen) [reply drivers/can/can_sja1000.c:641]: Put in another way, the TWAI module present in the ESP32, ESP32-C3 and others are not the same device - and will require a different DTS `compatible` along with a new front-end driver. Most of the shared SJA1000 functionality still applies, though.

File: drivers/can/can_sja1000.c
Comment (martinjaeger) [inline drivers/can/can_sja1000.c:671]:
```suggestion
		err = can_calc_prescaler(dev, &timing, config->bitrate);
```

**Review by martinjaeger — CHANGES_REQUESTED** [review]
Thanks for this beautiful piece of code. Made it really hard to spot anything to complain about, but I finally found something ;)

I don't have an ESP32 board at hand and the ESP32-C3 won't work yet (as discussed above), so I couldn't test in hardware myself. I did not double-check all register values in the header vs. the datasheet, but wrong registers should be covered by the tests.

**Review by uLipe — DISMISSED** [review]
Besides the intr_alloc comment above, looks great.

I won't block anyway, since other esp32 drivers don't check properly the interrupt allocator return value.

@sylvioalves do you know what is the reason behind to check interrupt allocator return value for xtensa ESP32? For C3 seems to be okay to me, since it will always succeed.

**Review by str4t0m — CHANGES_REQUESTED** [review]

File: drivers/can/can_sja1000.c
Comment (sylvioalves) [inline drivers/can/can_sja1000.c:126]: Is it necessary?

Comment (uLipe) [reply drivers/can/can_sja1000.c:126]: It seems okay, no? The mutex in this place is already held

Comment (uLipe) [reply drivers/can/can_sja1000.c:126]: Ahh now I get it, the unlock is right below.

Well for consistency I would keep, the compiler is smart enough to optimize here.

File: tests/drivers/can/timing/src/main.c
Comment (sylvioalves) [reply tests/drivers/can/timing/src/main.c:42]: Yes: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/twai.html

**Comment by sylvioalves** [comment]
@henrikbrixandersen LGTM besides pending comments. I think we can add support to ESP32S2 and ESP32C3 TWAI in another PR?

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen LGTM besides pending comments. I think we can add support to ESP32S2 and ESP32C3 TWAI in another PR?

Thank you for reviewing. Yes, support for ESP32-S2 and -C3 can be added in follow-up PRs.

File: drivers/can/can_sja1000.c
Comment (henrikbrixandersen) [reply drivers/can/can_sja1000.c]: Thanks, removed.

File: drivers/can/can_sja1000.c
Comment (henrikbrixandersen) [reply drivers/can/can_sja1000.c:698]: Thanks, fixed.

File: drivers/can/can_sja1000.c
Comment (henrikbrixandersen) [reply drivers/can/can_sja1000.c]: Thank you @martinjaeger :-) Added the missing return value check.

File: tests/drivers/can/timing/src/main.c
Comment (henrikbrixandersen) [reply tests/drivers/can/timing/src/main.c]: Fixed the comment, thanks.

**Comment by henrikbrixandersen** [comment]
@str4t0m, @martinjaeger, @sylvioalves, and @uLipe: Please re-review.

**Review by sylvioalves — DISMISSED** [review]

**Review by uLipe — DISMISSED** [review]
LGTM

**Comment by henrikbrixandersen** [comment]
Rebased to address the unrelated CI issue.

**Comment by henrikbrixandersen** [comment]
CI is green now. No changes to the contents of this PR. Please re-approve.

**Review by uLipe — APPROVED** [review]

**Review by str4t0m — APPROVED** [review]

**Review by martinjaeger — APPROVED** [review]

**Pull request merged**
